engine/src/main/java/org/hibernate/search/engine/mapper/mapping/building/impl/IndexSchemaFilter.java

```diff
@@ -196,24 +195,91 @@ private String getPathFromSameIndexedEmbeddedSinceNoCompositionLimits(IndexedEmb
 		if ( hasCompositionLimits() ) {
 			return null;
 		}
-		else if ( parent != null ) {
-			if ( this.definition.equals( definition ) ) {
-				// Same IndexedEmbedded as the one passed as a parameter
-				return this.definition.relativePrefix();
-			}
-			else {
-				String path = parent.getPathFromSameIndexedEmbeddedSinceNoCompositionLimits( definition );
-				return path == null ? null : path + this.definition.relativePrefix();
+		if ( this.definition == null ) {
+			// we are at root
+			return null;
+		}
+
+		return findCycle( definition );
+	}
+
+	/*
+	 * The idea is to find the longest cycle so far possible.
+	 * There is a chance that the cycle may repeat a few times if there are exclude paths that may break the cycle at some point.
+	 * Let's assume we have a cycle where we are trying to add next `a.` definition same as the already added ones before:
+	 * a.b.c.a.b.c.a.b.c.a.b.c. + a.
+	 * 1   ^ 2     3     4        5
+	 *
+	 * - we'll be looking for a definition `a.` with index `1` and that would define our "longest cycle".
+	 * - we want to know the filter representing the definition `c.` (marked with `^`) that is right before the `a.` with index `2` as well as the
+	 * cycle-path leading to it from `5` -- "a.b.c.a.b.c.a.b.c."
+	 * that will be the filter and the path we'll pass to the "isPotentiallyExcludedPath" check. These will be our `previousCycleFilter` and `previousCycle`
```

Comment:
```suggestion
	 * that will be the filter and the path we'll pass to the "isPotentiallyExcludedPath" check. These will be our `previousCycleFilter` and `pathFromPreviousCycleFilter`
```

You probably should have another look at the rest of this comment to be sure it still matches the code.